### PR TITLE
`Programming exercises`: Add form status bar to update view

### LINKED ISSUE #8024: Clone repository button on exercise detail view does not work
### Describe the bug

When viewing the programming exercise details as an instructor, the clone repository button has no content:
grafik

### To Reproduce

1. Navigate to Course Management -> Exercises -> Exercise Details
2. Click on "Clone Repository" (next to one of the 3 repositories)
3. See that the content is empty

### Expected behavior

The repository link should be displayed.

### Screenshots

_No response_

### Which version of Artemis are you seeing the problem on?

6.8.0

### What browsers are you seeing the problem on?

Firefox

### Additional context

_No response_

### Relevant log output

```shell
ERROR TypeError: null has no properties
    Do detail-overview-list.component.html:9
```

### LINKED ISSUE #8027: `Programming exercises`: Diff is not working
### Describe the bug

The diff is not working in programming exercises. 

### To Reproduce

Open any programming exercise and click in the 'show' button to see the diff between the template end the solution. 

### Expected behavior

Should show the diff. 

### Screenshots

image

image

### Which version of Artemis are you seeing the problem on?

6.8.0

### What browsers are you seeing the problem on?

Chrome, Safari

### Additional context

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #8033: `Programming exercises`: The status message does not disappear when importing a programming exercise
### Describe the bug

The status message does not disappear when importing a programming exercise. Even pressing X (close). 

### To Reproduce

1. Import a programming exercise and verify that the status message does not disappear when importing a programming exercise. Even pressing X (close). 
2. The same occurs when updating the exercise.

See screenshots below. 

### Expected behavior

The status message should disappear (after some seconds) when importing a programming exercise. Also when pressing X (close). 

### Screenshots

Screenshot 2024-02-14 at 14 42 17

image

### Which version of Artemis are you seeing the problem on?

6.8.0

### What browsers are you seeing the problem on?

Chrome, Safari

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by milljoniaer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Bitbucket** and **Bamboo**.

### Motivation and Context
This PR adds the validation bar to programming exercises and removes the programming-exercise-wizard, as its functionality is replaced by the bar.

also closes #8024 , closes #8033 , closes #8027 and the bug that the consistency check did not work properly by fixing a type error

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Create/Import a Programming Exercise
3. Use different settings/configuration
4. Verify that the validation bar always reflects the state of the form

#### Exam Mode Testing
Same for an exam programming exercise 

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise-editable-instruction.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/instructions-editor/programming-exercise-editable-instruction.component.ts.html) | 87.06% | ✅ |
| [programming-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/programming-exercise-update.component.ts.html) | 88.32% | ✅ |
| [remove-auxiliary-repository-button.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/remove-auxiliary-repository-button.component.ts.html) | 100% | ✅ |
| [programming-exercise-difficulty.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.ts.html) | 100% | ✅ |
| [programming-exercise-grading.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-grading.component.ts.html) | 96.66% | ✅ |
| [programming-exercise-information.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-information.component.ts.html) | 93.75% | ✅ |
| [programming-exercise-language.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-language.component.ts.html) | 100% | ✅ |
| [programming-exercise-problem.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-problem.component.ts.html) | 100% | ✅ |
| [programming-exercise-lifecycle.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts.html) | 93.02% | ✅ |
| [programming-exercise-test-schedule-date-picker.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/lifecycle/programming-exercise-test-schedule-date-picker.component.ts.html) | 94.73% | ✅ |
| [table-editable-field.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6079/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/table/table-editable-field.component.ts.html) | 100% | ✅ |

### Screenshots

New edit mode:

https://github.com/ls1intum/Artemis/assets/78978542/19a2f114-8431-477d-833b-170194f7e09b

Old wizard mode, which is now combined with the normal mode to the new mode:

https://github.com/ls1intum/Artemis/assets/78978542/3940e6e6-6c03-48d2-b63d-986e2e7f6ba6

## Summary by CodeRabbit

- **New Features**
	- Added dynamic form status updates and validation across various exercise management components.
	- Introduced new form elements and conditional rendering in exercise update components to enhance user experience.
	- Implemented event emitters for real-time updates in programming exercises.

- **Improvements**
	- Updated conditional checks and rendering logic in templates for clarity and efficiency.
	- Reorganized HTML template structure for better readability and maintenance.
	- Enhanced accessibility and localization support with new `id` attributes for translation.

- **Bug Fixes**
	- Corrected the icon for the `RemoveAuxiliaryRepositoryButtonComponent` to better reflect its functionality.

- **Tests**
	- Added Cypress and JavaScript test cases for new features and modifications, ensuring robust functionality.
	- Improved test coverage for programming exercise components and settings.

- **Refactor**
	- Consolidated logic for validating exercise inputs and managing subscriptions for dynamic form sections.

### CODE REVIEW

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.html [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
-        }
-        @if (!fileUploadExercise.id) {
+        } @else if (!fileUploadExercise.id) {
             <h2 id="jhi-file-upload-exercise-heading-create" jhiTranslate="artemisApp.fileUploadExercise.home.createLabel">Create a new File Upload Exercise</h2>
-        }
-        @if (isImport) {
+        } @else if (isImport) {
             <h2 id="jhi-file-upload-exercise-heading-import" jhiTranslate="artemisApp.fileUploadExercise.home.importLabel">Import a File Upload Exercise</h2>
         }
         <jhi-documentation-button [type]="documentationType"></jhi-documentation-button>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.html:10]:
```suggestion
        <jhi-documentation-button [type]="documentationType" />
```

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise-update.component.html [hunk 2]
```diff
@@ -1,16 +1,12 @@
...
-        </div>
+    <div class="d-flex align-items-center gap-2">
+        @if (!modelingExercise.id) {
+            <h2 id="jhi-modeling-exercise-heading-create" jhiTranslate="artemisApp.modelingExercise.home.createLabel">Create a new Modeling Exercise</h2>
+        } @else if (!isImport && modelingExercise.id) {
+            <h2 id="jhi-modeling-exercise-heading-edit" jhiTranslate="artemisApp.modelingExercise.home.editLabel">Edit a Modeling Exercise</h2>
+        } @else if (isImport) {
+            <h2 id="jhi-modeling-exercise-heading-import" jhiTranslate="artemisApp.modelingExercise.home.importLabel">Import Modeling Exercise</h2>
+        }
         <jhi-documentation-button [type]="documentationType"></jhi-documentation-button>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/modeling/manage/modeling-exercise-update.component.html:10]:
```suggestion
        <jhi-documentation-button [type]="documentationType" />
```

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-grading.component.html [hunk 3]
```diff
@@ -93,6 +95,7 @@ <h3 jhiTranslate="artemisApp.programmingExercise.wizardMode.detailedSteps.gradin
             #submissionPolicyForm
             [editable]="!programmingExerciseCreationConfig.isEdit"
             [programmingExercise]="programmingExercise"
+            (submissionPolicyTypeChange)="calculateFormStatus()"
         ></jhi-submission-policy-update>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-grading.component.html:99]:
```suggestion
        />
```

**Review by AntonGeTUM — APPROVED** [review]
Tested on ts11, works as shown in the video.

(Btw, when selecting dates for the exercise timeline, the status bar disappears while the calendar is shown, just a bit odd.)

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-problem.component.html [hunk 4]
```diff
@@ -31,6 +33,7 @@ <h3 jhiTranslate="artemisApp.programmingExercise.wizardMode.detailedSteps.proble
                 [showStatus]="!!programmingExercise?.id"
                 [editMode]="!!programmingExercise?.id"
                 (hasUnsavedChanges)="programmingExerciseCreationConfig.hasUnsavedChanges = $event"
+                (instructionChange)="problemStatementChange.emit($event)"
                 [forceRender]="programmingExerciseCreationConfig.rerenderSubject"
                 class="form__editable-instructions"
             ></jhi-programming-exercise-editable-instructions>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-problem.component.html:39]:
```suggestion
            />
```

File: src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-update.component.html [hunk 5]
```diff
@@ -1,12 +1,10 @@
...
-        }
-        @if (!isImport && textExercise.id) {
+        } @else if (!isImport && textExercise.id) {
             <h2 id="jhi-text-exercise-heading-edit" jhiTranslate="artemisApp.textExercise.home.editLabel">Edit Text Exercise</h2>
-        }
-        @if (isImport) {
+        } @else if (isImport) {
             <h2 id="jhi-text-exercise-heading-import" jhiTranslate="artemisApp.textExercise.home.importLabel">Import Text Exercise</h2>
         }
         <jhi-documentation-button [type]="documentationType"></jhi-documentation-button>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-update.component.html:10]:
```suggestion
        <jhi-documentation-button [type]="documentationType" />
```

**Review by MaximilianAnzinger — COMMENTED** [review]
Overall code looks good 👍
Left some small suggestions.

**Review by Gusti2010 — APPROVED** [review]
Manually tested on ts11_Legacy. Works as expected!

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by rstief — DISMISSED** [review]
Code looks good 👍

**Review by nata-81 — APPROVED** [review]
Tested on ts4 works as expected.
Except that if the release and/or start dates aren't set, but the due date is, the grading still is displayed as green, but I actually would expect it to be gray. But maybe this is intended to be like that.
image

**Review by rstief — DISMISSED** [review]
reapprove

**Review by AntonGeTUM — APPROVED** [review]

**Review by rstief — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by nata-81 — APPROVED** [review]
Retested on TS4, it works as expected.

**Review by rabeatwork — APPROVED** [review]

**Pull request merged**